### DISCUSSION TITLE
How do you stop Javascript from a <script> block which has been removed after an AJAX call?

### DISCUSSION THREAD

**Discussion by josephernest** [unpoly/unpoly#409] [General]
Let's say I click on a `<a href...` link, unpoly intercepts it and does an AJAX call, then replaces a part of the DOM, for example `#container`'s innerHTML.

Let's say this new HTML code contains a `<script>setInterval(function () { alert(); }, 5000);</script>` block. It will be executed.

Then you click on *another* link, and `#container` is now replaced again by something else. This `<script>` is no more there in the new HTML content.

**Question: do you stop the setInterval from running? If so, how?**

Thanks @triskweline!

**Comment by triskweline** [comment]
Move your code into a [compiler](https://unpoly.com/up.compiler). Compilers can return a [destructor function](https://unpoly.com/up.compiler#cleaning-up-after-yourself) that runs when the element is removed from the DOM.

**Reply by josephernest** [reply]
Thanks @triskweline. 

> If your compiler returns a function, Unpoly will use this as a destructor to clean up if the element leaves the DOM

Out of curiosity, how do you detect if an element is currently being destroyed? There can be many cases: because the `<div class="screen2">` is replaced by *another* `<div>` if the user clicks on a link with `up-target=".screen2"`, or also because the user did "history back" and this `<div>` disappears, and probably also other cases...

In which Unpoly source file do you deal with all these cases? I looked in the repo, but not sure to find where exactly this magic happens :)

**Reply by josephernest** [reply]
Do you maybe use `MutationObserver` HTML5 API to detect changes in the DOM? (I didn't find this in the code, but maybe I searched in the wrong places!).

**Reply by triskweline** [reply]
> Out of curiosity, how do you detect if an element is currently being destroyed?

When Unpoly swaps a fragment, it already knows that its being remove and runs destructors. If you manually remove an element without Unpoly, you should use `up.destroy()` to ensure destructors are run.

> Do you maybe use MutationObserver HTML5 API to detect changes in the DOM? 

Unpoly may use `MutationObserver` in the future so people that do a lot of manual DOM manipulation don't need to use `up.destroy()` and `up.hello()`. We would need to run performance tests since MutationObserver records a lot of change events.

**Comment by josephernest** [comment]
Thanks @triskweline. By default, if we don't use a `compiler`, it doesn't insert the `<script>` or it doesn't run it?

`index.html`

```
<!doctype html>
<html>
<head>
<script src="https://unpkg.com/unpoly@2.6.1/unpoly.min.js"></script>
</head>
<body>
<a href="front.html" class="one" up-target=".one">A1</a>
<a href="front.html" class="two" up-target=".two">B1</a>
</body>
</html>
```

`front.html`

```
<!doctype html>
<html>
<head>
<script src="https://unpkg.com/unpoly@2.6.1/unpoly.min.js"></script>
</head>
<body>
<a href="index.html" class="one" up-target=".one">A2</a>
<a href="index.html" class="two" up-target=".two">B2<script>setInterval(alert, 5000);</script></a>
</body>
</html>
```

Here this `setInterval` is never executed ^.

**Reply by triskweline** [reply]
By default `<script>` elements are not inserted, but you can enable this with [`up.fragment.config.runScripts = true`](https://unpoly.com/up.fragment.config#config.runScripts).
